File: go/authoring/authoring.go
// SPDX-License-Identifier: Apache-2.0

// Package authoring renders a declarative.Plan into a versioned,
// reversible Keystone migration bundle: an `NNN_name.up.sql` /
// `NNN_name.down.sql` pair that satisfies the five migration-authoring
// standards (statement_timeout header, leading comment, schema-relative
// DDL, idempotent IF EXISTS rollbacks, paired up/down) by construction.
//
// It is the bridge between schema-as-code and the GitOps bundle format:
// `keystonectl migrate diff` and `keystonectl scaffold` both call
// RenderUpDown to turn a computed diff into files an author commits. The
// package is pure — no database, no filesystem, no Kubernetes — so it is
// trivially unit-testable and reusable by any tool that already has a
// *declarative.Plan in hand.
//
// The forward statements come from plan.Statements verbatim; the down
// file replays plan.ReverseStatements in reverse order (so a rollback
// undoes the most-recent change first). Statements the differ marked
// irreversible (empty reverse) are rendered as explicit comment markers
// rather than silently dropped, keeping the down file an honest record
// of what can and cannot be rolled back.
package authoring

import (
	"fmt"
	"path/filepath"
	"regexp"
	"strings"
	"sync"

	"github.com/dogukanturhal/keystone-sdk/go/declarative"
	pg "github.com/dogukanturhal/keystone-sdk/go/sqlquote"
)

// DefaultStatementTimeout is prepended to every rendered file as
// `SET LOCAL statement_timeout = '<this>';` per migration-authoring §2.
// 30s fails fast on an accidental non-CONCURRENTLY index build against a
// large table rather than blasting a production lock.
const DefaultStatementTimeout = "30s"

// Options controls how a Plan is rendered into a bundle.
type Options struct {
	// Name is the human-readable change slug (e.g. "add_users_email_index").
	// Sanitised into the filename; non [a-z0-9_] runes become underscores.
	Name string

	// Version is the zero-padded ordinal prefix (e.g. "002"). Use
	// NextVersion to derive it from an existing directory.
	Version string

	// Schema is the target schema name, recorded in the file header for
	// reviewers. Does not affect the SQL (migrations are schema-relative;
	// the runner sets search_path).
	Schema string

	// StatementTimeout overrides DefaultStatementTimeout when non-empty.
	StatementTimeout string

	// GeneratedBy is a provenance string for the header comment
	// (e.g. "keystonectl v0.2.0").
	GeneratedBy string

	// StripSchema, when non-empty, rewrites the generated SQL to be
	// schema-relative by removing the `"<StripSchema>".` qualifier the
	// differ emits (it qualifies every object with observed.Schema). This
	// yields portable migrations the runner can apply against any target
	// schema via search_path, per migration-authoring §4. Set it to the
	// schema the diff was computed against.
	StripSchema string
}

// Bundle is the rendered file set for a single migration.
type Bundle struct {
	Version  string // "002"
	Name     string // sanitised slug, "add_users_email_index"
	UpFile   string // "002_add_users_email_index.up.sql"
	DownFile string // "002_add_users_email_index.down.sql"
	Up       string // up.sql contents
	Down     string // down.sql contents
}

// Files returns the bundle as a filename→content map suitable for
// writing to disk and for migration.BuildSum (which the caller uses to
// regenerate keystone.sum over the whole directory).
func (b Bundle) Files() map[string]string {
	return map[string]string{
		b.UpFile:   b.Up,
		b.DownFile: b.Down,
	}
}

// RenderUpDown renders plan into a Bundle. The forward file applies
// plan.Statements in order; the down file applies plan.ReverseStatements
// in reverse order so rollback unwinds most-recent-first. A nil plan
// yields empty Up/Down bodies (header only).
func RenderUpDown(plan *declarative.Plan, opts Options) Bundle {
	timeout := opts.StatementTimeout
	if timeout == "" {
		timeout = DefaultStatementTimeout
	}
	slug := sanitizeSlug(opts.Name)
	base := opts.Version
	if base != "" && slug != "" {
		base = base + "_" + slug
	} else if slug != "" {
		base = slug
	}

	b := Bundle{
		Version:  opts.Version,
		Name:     slug,
		UpFile:   base + ".up.sql",
		DownFile: base + ".down.sql",
	}

	var up, down []entry
	if plan != nil {
		for _, s := range plan.Statements {
			if s = strings.TrimSpace(stripSchema(s, opts.StripSchema)); s != "" {
				up = append(up, entry{sql: s})
			}
		}
		// Reverse order: ReverseStatements[i] undoes Statements[i], so a
		// rollback must run the LAST change's reverse first.
		for i := len(plan.ReverseStatements) - 1; i >= 0; i-- {
			r := strings.TrimSpace(stripSchema(plan.ReverseStatements[i], opts.StripSchema))
			if r == "" {
				// Irreversible forward op (e.g. DROP COLUMN data loss).
				fwd := ""
				if i < len(plan.Statements) {
					fwd = firstLine(plan.Statements[i])
				}
				down = append(down, entry{comment: "irreversible: no automatic rollback for: " + fwd})
				continue
			}
			down = append(down, entry{sql: r})
		}
	}

	// A down file with no executable statements (fully irreversible, or an
	// empty plan) still needs a valid, non-empty body so the rollback
	// orchestrator does not fail on a missing/empty file (§5).
	if !hasSQL(down) {
		down = append(down, entry{
			sql: "SELECT 'irreversible: see migration notes for the rollback procedure' AS rollback_note",
		})
	}

	b.Up = renderFile(opts, slug, "up", timeout, up)
	b.Down = renderFile(opts, slug, "down", timeout, down)
	return b
}

// NextVersion returns the next zero-padded ordinal given the names of
// existing migrations (filenames or bare version strings). It reads the
// leading run of digits of each, takes the max, and increments — padding
// to at least three digits, or wider if any existing version already is.
// An empty input yields "001".
func NextVersion(existing []string) string {
	maxN, width := 0, 3
	for _, e := range existing {
		b := filepath.Base(e)
		j := 0
		for j < len(b) && b[j] >= '0' && b[j] <= '9' {
			j++
		}
		if j == 0 {
			continue
		}
		digits := b[:j]
		n := 0
		for _, c := range digits {
			n = n*10 + int(c-'0')
		}
		if len(digits) > width {
			width = len(digits)
		}
		if n > maxN {
			maxN = n
		}
	}
	return fmt.Sprintf("%0*d", width, maxN+1)
}

// entry is one line of a rendered file: either a SQL statement (gets a
// trailing semicolon) or a standalone comment (emitted verbatim).
type entry struct {
	sql     string
	comment string
}

func hasSQL(entries []entry) bool {
	for _, e := range entries {
		if e.sql != "" {
			return true
		}
	}
	return false
}

// renderFile assembles the file body: provenance comment header, the
// mandatory SET LOCAL statement_timeout, then the statements/comments
// separated by blank lines.
func renderFile(opts Options, slug, direction, timeout string, entries []entry) string {
	var sb strings.Builder

	// Leading comment (satisfies require-migration-comment; gives
	// reviewers intent + provenance at a glance).
	title := slug
	if title == "" {
		title = "migration"
	}
	fmt.Fprintf(&sb, "-- %s (%s)\n", title, direction)
	gen := opts.GeneratedBy
	if gen == "" {
		gen = "keystone authoring"
	}
	fmt.Fprintf(&sb, "-- generated by %s from a declarative schema diff; edit the SchemaDefinition, not this file.\n", gen)
	if opts.Schema != "" {
		fmt.Fprintf(&sb, "-- target schema: %s\n", opts.Schema)
	}
	sb.WriteByte('\n')

	// Mandatory per-file statement timeout (migration-authoring §2).
	fmt.Fprintf(&sb, "SET LOCAL statement_timeout = '%s';\n", timeout)

	for _, e := range entries {
		sb.WriteByte('\n')
		if e.comment != "" {
			fmt.Fprintf(&sb, "-- %s\n", e.comment)
			continue
		}
		sb.WriteString(e.sql)
		// The differ emits bare statements without a terminator; add one.
		if !strings.HasSuffix(strings.TrimRight(e.sql, " \t"), ";") {
			sb.WriteByte(';')
		}
		sb.WriteByte('\n')
	}
	return sb.String()
}

// sanitizeSlug lowercases and replaces any rune outside [a-z0-9_] with an
// underscore, collapsing runs and trimming leading/trailing underscores.
func sanitizeSlug(s string) string {
	s = strings.ToLower(strings.TrimSpace(s))
	var sb strings.Builder
	prevUnderscore := false
	for _, r := range s {
		switch {
		case (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9'):
			sb.WriteRune(r)
			prevUnderscore = false
		default:
			if !prevUnderscore {
				sb.WriteByte('_')
				prevUnderscore = true
			}
		}
	}
	return strings.Trim(sb.String(), "_")
}

// stripSchema removes the `"<schema>".` qualifier the differ prepends to
// every object, yielding schema-relative DDL. A no-op when schema is
// empty. The differ always quotes the schema (pg.QuoteIdentifier), so the
// token is unambiguous and will not collide with unqualified identifiers.
// StripSchemaQualifier removes the `"<schema>".` qualifier from a single
// statement, yielding schema-relative SQL.
//
// Exported for exec:// schema providers (cmd/keystone-gorm and friends),
// which render DDL through the differ and must hand Keystone unqualified
// statements: the dev-database normaliser applies them under a scratch
// schema's search_path, and a qualifier would pin every object to a
// schema that is not the one being diffed. Shares the one implementation
// with RenderUpDown so authored migrations and provider output can never
// disagree about what "schema-relative" means.
func StripSchemaQualifier(sql, schema string) string { return stripSchema(sql, schema) }

func stripSchema(sql, schema string) string {
	if schema == "" || sql == "" {
		return sql
	}
	// Quoted form: what the differ emits, since it quotes every
	// identifier it renders.
	sql = strings.ReplaceAll(sql, pg.QuoteIdentifier(schema)+".", "")

	// Unquoted form: what PostgreSQL's own deparsers emit. A view body
	// comes back from pg_get_viewdef as `SELECT ... FROM app.orders`,
	// with no quotes on a lowercase identifier, so the replacement above
	// misses it entirely and the qualifier is written into the authored
	// migration. Applying that bundle anywhere the authoring schema does
	// not exist fails with 42P01, and applying it where the schema does
	// exist is worse — the statement silently reads and writes the wrong
	// schema instead of the one the bundle was fanned out to.
	//
	// The boundary check keeps a schema named `app` from also rewriting
	// `myapp.` or a qualified `other.app.` reference. Go's regexp has no
	// lookbehind, so the preceding character is captured and put back.
	return unquotedSchemaRef(schema).ReplaceAllString(sql, "$1")
}

// unquotedSchemaRefCache memoises the per-schema pattern; stripSchema is
// called once per statement per rendered file.
var (
	unquotedSchemaRefCache sync.Map // schema string -> *regexp.Regexp
)

func unquotedSchemaRef(schema string) *regexp.Regexp {
	if re, ok := unquotedSchemaRefCache.Load(schema); ok {
		return re.(*regexp.Regexp)
	}
	re := regexp.MustCompile(`(^|[^A-Za-z0-9_."])` + regexp.QuoteMeta(schema) + `\.`)
	unquotedSchemaRefCache.Store(schema, re)
	return re
}

func firstLine(s string) string {
	s = strings.TrimSpace(s)
	if i := strings.IndexByte(s, '\n'); i >= 0 {
		return strings.TrimSpace(s[:i])
	}
	return s
}

File: rust/keystone-sdk/src/authoring.rs
// SPDX-License-Identifier: Apache-2.0

//! Renders a [`crate::declarative::Plan`] into a versioned, reversible
//! migration bundle (`NNN_name.up.sql` / `NNN_name.down.sql`).
//!
//! Ported from the Go `authoring` package. Pure — no database, filesystem,
//! or Kubernetes. The forward file applies `plan.statements` in order; the
//! down file replays `plan.reverse_statements` in REVERSE order (rollback
//! unwinds most-recent-first). Irreversible ops (empty reverse) are rendered
//! as explicit comment markers, not silently dropped.

use std::collections::BTreeMap;

use crate::declarative::Plan;
use crate::ident::quote_identifier;

/// Prepended to every rendered file as `SET LOCAL statement_timeout =
/// '<this>';`. 30s fails fast on an accidental non-CONCURRENTLY index build.
pub const DEFAULT_STATEMENT_TIMEOUT: &str = "30s";

/// Controls how a [`Plan`] is rendered into a [`Bundle`].
#[derive(Debug, Clone, Default)]
pub struct Options {
    /// Human-readable change slug (sanitised into the filename).
    pub name: String,
    /// Zero-padded ordinal prefix (e.g. `"002"`). See [`next_version`].
    pub version: String,
    /// Target schema name, recorded in the header (does not affect SQL).
    pub schema: String,
    /// Overrides [`DEFAULT_STATEMENT_TIMEOUT`] when non-empty.
    pub statement_timeout: String,
    /// Provenance string for the header comment.
    pub generated_by: String,
    /// When non-empty, strips the `"<strip_schema>".` qualifier the differ
    /// emits, yielding schema-relative (portable) DDL.
    pub strip_schema: String,
}

/// The rendered file set for a single migration.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct Bundle {
    pub version: String,
    pub name: String,
    pub up_file: String,
    pub down_file: String,
    pub up: String,
    pub down: String,
}

impl Bundle {
    /// The bundle as a filename→content map (for writing to disk + feeding
    /// `keystone.sum` regeneration).
    pub fn files(&self) -> BTreeMap<String, String> {
        let mut m = BTreeMap::new();
        m.insert(self.up_file.clone(), self.up.clone());
        m.insert(self.down_file.clone(), self.down.clone());
        m
    }
}

/// One line of a rendered file: a SQL statement (gets a trailing `;`) or a
/// standalone comment (emitted verbatim). Exactly one field is non-empty.
struct Entry {
    sql: String,
    comment: String,
}

fn has_sql(entries: &[Entry]) -> bool {
    entries.iter().any(|e| !e.sql.is_empty())
}

/// Renders `plan` into a [`Bundle`]. `None` yields header-only bodies.
pub fn render_up_down(plan: Option<&Plan>, opts: &Options) -> Bundle {
    let timeout = if opts.statement_timeout.is_empty() {
        DEFAULT_STATEMENT_TIMEOUT
    } else {
        &opts.statement_timeout
    };
    let slug = sanitize_slug(&opts.name);
    let base = if !opts.version.is_empty() && !slug.is_empty() {
        format!("{}_{}", opts.version, slug)
    } else if !slug.is_empty() {
        slug.clone()
    } else {
        opts.version.clone()
    };

    let mut up: Vec<Entry> = Vec::new();
    let mut down: Vec<Entry> = Vec::new();
    if let Some(plan) = plan {
        for s in &plan.statements {
            let s = strip_schema(s, &opts.strip_schema);
            let s = s.trim();
            if !s.is_empty() {
                up.push(Entry {
                    sql: s.to_string(),
                    comment: String::new(),
                });
            }
        }
        // Reverse order: reverse_statements[i] undoes statements[i].
        for i in (0..plan.reverse_statements.len()).rev() {
            let r = strip_schema(&plan.reverse_statements[i], &opts.strip_schema);
            let r = r.trim();
            if r.is_empty() {
                let fwd = if i < plan.statements.len() {
                    first_line(&plan.statements[i])
                } else {
                    String::new()
                };
                down.push(Entry {
                    sql: String::new(),
                    comment: format!("irreversible: no automatic rollback for: {fwd}"),
                });
                continue;
            }
            down.push(Entry {
                sql: r.to_string(),
                comment: String::new(),
            });
        }
    }

    if !has_sql(&down) {
        down.push(Entry {
            sql: "SELECT 'irreversible: see migration notes for the rollback procedure' AS rollback_note".to_string(),
            comment: String::new(),
        });
    }

    Bundle {
        version: opts.version.clone(),
        name: slug.clone(),
        up_file: format!("{base}.up.sql"),
        down_file: format!("{base}.down.sql"),
        up: render_file(opts, &slug, "up", timeout, &up),
        down: render_file(opts, &slug, "down", timeout, &down),
    }
}

/// Returns the next zero-padded ordinal given existing migration names
/// (filenames or bare versions). Reads each leading digit run, takes the
/// max + 1, padded to ≥3 digits (wider if any existing version is). Empty
/// input yields `"001"`.
pub fn next_version(existing: &[String]) -> String {
    let mut max_n = 0u64;
    let mut width = 3usize;
    for e in existing {
        let b = e.rsplit('/').next().unwrap_or(e);
        let digits: String = b.chars().take_while(|c| c.is_ascii_digit()).collect();
        if digits.is_empty() {
            continue;
        }
        let n: u64 = digits.parse().unwrap_or(0);
        if digits.len() > width {
            width = digits.len();
        }
        if n > max_n {
            max_n = n;
        }
    }
    format!("{:0width$}", max_n + 1, width = width)
}

fn render_file(
    opts: &Options,
    slug: &str,
    direction: &str,
    timeout: &str,
    entries: &[Entry],
) -> String {
    let mut sb = String::new();

    let title = if slug.is_empty() { "migration" } else { slug };
    sb.push_str(&format!("-- {title} ({direction})\n"));
    let gen = if opts.generated_by.is_empty() {
        "keystone authoring"
    } else {
        &opts.generated_by
    };
    sb.push_str(&format!(
        "-- generated by {gen} from a declarative schema diff; edit the SchemaDefinition, not this file.\n"
    ));
    if !opts.schema.is_empty() {
        sb.push_str(&format!("-- target schema: {}\n", opts.schema));
    }
    sb.push('\n');

    sb.push_str(&format!("SET LOCAL statement_timeout = '{timeout}';\n"));

    for e in entries {
        sb.push('\n');
        if !e.comment.is_empty() {
            sb.push_str(&format!("-- {}\n", e.comment));
            continue;
        }
        sb.push_str(&e.sql);
        if !e.sql.trim_end_matches([' ', '\t']).ends_with(';') {
            sb.push(';');
        }
        sb.push('\n');
    }
    sb
}

/// Lowercases and replaces any char outside `[a-z0-9_]` with a single
/// underscore (collapsing runs), trimming leading/trailing underscores.
fn sanitize_slug(s: &str) -> String {
    let s = s.trim().to_lowercase();
    let mut out = String::new();
    let mut prev_underscore = false;
    for c in s.chars() {
        if c.is_ascii_lowercase() || c.is_ascii_digit() {
            out.push(c);
            prev_underscore = false;
        } else if !prev_underscore {
            out.push('_');
            prev_underscore = true;
        }
    }
    out.trim_matches('_').to_string()
}

/// Removes the `"<schema>".` qualifier the differ prepends to every object.
fn strip_schema(sql: &str, schema: &str) -> String {
    if schema.is_empty() || sql.is_empty() {
        return sql.to_string();
    }
    sql.replace(&format!("{}.", quote_identifier(schema)), "")
}

fn first_line(s: &str) -> String {
    let s = s.trim();
    match s.find('\n') {
        Some(i) => s[..i].trim().to_string(),
        None => s.to_string(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn next_version_cases() {
        assert_eq!(next_version(&[]), "001");
        assert_eq!(
            next_version(&["001_init.up.sql".into(), "002_foo.up.sql".into()]),
            "003"
        );
        assert_eq!(next_version(&["007".into()]), "008");
        // Width widens to match the widest existing version.
        assert_eq!(next_version(&["0001_x.up.sql".into()]), "0002");
        // Non-numeric entries are skipped.
        assert_eq!(
            next_version(&["readme.md".into(), "005_a.sql".into()]),
            "006"
        );
    }

    #[test]
    fn sanitize_slug_cases() {
        assert_eq!(
            sanitize_slug("Add Users Email Index"),
            "add_users_email_index"
        );
        assert_eq!(sanitize_slug("  --weird/Name--  "), "weird_name");
        assert_eq!(sanitize_slug("already_ok"), "already_ok");
    }

    #[test]
    fn render_up_down_basic() {
        let plan = Plan {
            statements: vec![
                "CREATE INDEX CONCURRENTLY IF NOT EXISTS \"app\".\"idx\" ON \"app\".\"t\" USING btree (\"c\")".into(),
                "ALTER TABLE \"app\".\"t\" DROP COLUMN IF EXISTS \"old\"".into(),
            ],
            reverse_statements: vec![
                "DROP INDEX CONCURRENTLY IF EXISTS \"app\".\"idx\"".into(),
                String::new(), // drop column = irreversible
            ],
            warnings: vec![],
            destructive_ops: 1,
        };
        let opts = Options {
            name: "add idx".into(),
            version: "002".into(),
            schema: "app".into(),
            strip_schema: "app".into(),
            generated_by: "keystonectl v0.2.0".into(),
            ..Default::default()
        };
        let b = render_up_down(Some(&plan), &opts);
        assert_eq!(b.up_file, "002_add_idx.up.sql");
        assert_eq!(b.down_file, "002_add_idx.down.sql");

        // Up: header + timeout + 2 statements, schema-stripped, terminated.
        assert!(b.up.starts_with("-- add_idx (up)\n"));
        assert!(b
            .up
            .contains("-- generated by keystonectl v0.2.0 from a declarative schema diff"));
        assert!(b.up.contains("-- target schema: app\n"));
        assert!(b.up.contains("SET LOCAL statement_timeout = '30s';\n"));
        assert!(b.up.contains(
            "CREATE INDEX CONCURRENTLY IF NOT EXISTS \"idx\" ON \"t\" USING btree (\"c\");\n"
        ));
        assert!(b
            .up
            .contains("ALTER TABLE \"t\" DROP COLUMN IF EXISTS \"old\";\n"));
        assert!(!b.up.contains("\"app\"."));

        // Down: reverse order — the DROP COLUMN's irreversible marker comes
        // first (it's the last forward op), then the index drop.
        let irr = b.down.find("-- irreversible: no automatic rollback for: ALTER TABLE \"app\".\"t\" DROP COLUMN IF EXISTS \"old\"").unwrap();
        let dropidx = b
            .down
            .find("DROP INDEX CONCURRENTLY IF EXISTS \"idx\";")
            .unwrap();
        assert!(
            irr < dropidx,
            "irreversible marker should precede the index drop"
        );
    }

    #[test]
    fn render_empty_plan_has_rollback_note() {
        let b = render_up_down(
            None,
            &Options {
                name: "noop".into(),
                version: "001".into(),
                ..Default::default()
            },
        );
        assert!(b.down.contains("rollback_note"));
        assert!(b.up.contains("SET LOCAL statement_timeout = '30s';"));
    }
}
